app/scripts/controllers/permissions/specifications.js

```diff
@@ -176,6 +195,53 @@ export const getPermissionSpecifications = ({
         }
       },
     },
+
+    [PermissionNames.wallet_switchEthereumChain]: {
+      permissionType: PermissionType.RestrictedMethod,
+      targetName: PermissionNames.wallet_switchEthereumChain,
+      allowedCaveats: [CaveatTypes.restrictNetworkSwitching],
+
+      factory: (permissionOptions, requestData) => {
+        if (Array.isArray(permissionOptions.caveats)) {
+          throw new Error(
+            `${PermissionNames.wallet_switchEthereumChain} error: Received unexpected caveats. Any permitted caveats will be added automatically.`,
+          );
+        }
+
+        // This value will be further validated as part of the caveat.
+        if (!requestData.approvedChainIds) {
+          throw new Error(
+            `${PermissionNames.wallet_switchEthereumChain} error: No approved networks specified.`,
+          );
+        }
+
+        const caveat = CaveatFactories[CaveatTypes.restrictNetworkSwitching](
+          requestData.approvedChainIds,
+        );
+
+        return constructPermission({
+          ...permissionOptions,
+          caveats: [caveat],
+        });
+      },
+      methodImplementation: async (_args) => {
+        // should we actually move the implementation here?
+        // This would make putting the permissioning behind a feature flag a bit more difficult?
+        // Interestingly the validator is still called when granting permissions
+      },
```

Comment:
@Gudahtt I'm realizing the new "permissioned" implementation should technically live here and then get passed in as hook to the `wallet-add-ethereum-chain` handler... I could do this but I think it will make the feature flag setup a bit more complex... WDYT?

Reply:
cc @rekmarks

Reply:
I'm hoping we go with Erik's idea to replace restricted method permissions with endowment permissions (https://github.com/MetaMask/core/issues/4238), which would mean we'd want to enforce it in the method handler, as you've done already in this PR. So this looks good to me.

I noticed that this is never executed because we have the permission middleware placed later in the RPC pipeline than our unrestricted "RPC Method middleware" methods. This seems a bit inappropriate because it would be too easy to add a permission and not notice that it's never enforced.

If/when we fix that problem, if we still have restricted method based permissions at that point, we'll have to move the implementation here or at least call it from here.

Reply:
I agree with what @Gudahtt said, although: is this a runtime configurable feature flag, or do we expect this feature to be on or off for the lifetime of the application? If the latter, we could just conditionally include this permission specification when the controller is initialized.

Reply:
Copying over conversations from slack for visibility:

**From me (@adonesky1):** 
_So I'm a bit confused about whether we can actually achieve the UX we want in the current architecture if we treat wallet_switchEthereumChain as a restricted method in the same way as we do eth_accounts .
What I mean by this is that when a user calls wallet_switchEthereumChain with a chainId for which they do not currently have permissions we want them to get a permissions confirmation to add that chain not simply to throw an error as we do in the case of eth_accounts . I was achieving this in [this PR](https://github.com/MetaMask/metamask-extension/pull/24415) by explicitly [laying out this logic in the wallet_switchEthereumChain handler](https://github.com/MetaMask/metamask-extension/pull/24415/files#diff-93f17dab5026baf840aec766ac5b6ebcd35d5bb0abb45e2c3f41791423f2a488R150-R179) itself.
If we move the restricted method method implementation into the specification file and have it hit in the permission-middleware (ahead of unrestriced method implemetnations with [Erik's PR](https://github.com/MetaMask/metamask-extension/pull/24472)) where executeRestrictedMethod will throw without any hooks to request the missing permissions I don't believe we can support this kind of flow without modifying the middleware or PermissionController. Does this make sense or am I missing something?
I suppose the alternative is to have wallet_switchEthereumChain be in the unrestrictedMethods list as with eth_requestAccounts but call to a this.permissionController.executeRestrictedMethod(RestrictedMethods.wallet_switchEthereumChain) hook in the wallet_switchEthereumHandler and then the restricted specification method implementation just implements the actual switching logic, not the permissions checks or input validation?_

from @rekmarks :

_Ah, yeah. We are actually already in this situation (restricted method actually implemented elsewhere) with eth_accounts, which of course returns an empty array and not an unauthorized error if no permission is found.

I think the only place we actually use restricted methods as intended is in / for Snaps.

Okay, I think we have three options:
1. Implement as restricted method:
a. with dummy implementation (just throw an error), then never call it
b. with actual implementation, that you actually call through executeRestrictedMethod()
2. Implement as endowment permission, and handle enforcement directly in method implementation (using RPC handler hooks)

The thing I don’t like about 2 is that it leaves us in this awkward state between restricted methods and endowment permissions.
But it’s definitely the most direct route, options 1a and b are circuitous._

from @adonesky1 :

_That's basically what I'd done in the PR (prior to rebasing on yours) in every way except that the permission type is called restricted but then its also in the unrestrictedMethods array and so allowed to pass through to the handler since theres nothing enforcing that restricted and unrestricted methods are mutually exclusive :sweat_smile: 

And so I guess yeah lets make it an endowment?_

Reply:
So we've decided to make it an endowment permission, which is done here: https://github.com/MetaMask/metamask-extension/pull/24415/commits/1a25e0f43e2813f5f08d215a49dcc2668deb1b7d